docs/assembly.rst

```diff
@@ -1,17 +1,195 @@
+    Case =
+        'case' Expression ':' Block
+    ForLoop =
+        'for' Block Expression Block Block
+    BreakContinue =
+        'break' | 'continue'
+    SubAssembly =
+        'assembly' Identifier Block
+    FunctionCall =
+        Identifier '(' ( Expression ( ',' Expression )* )? ')'
+    IdentifierOrList = Identifier | '(' IdentifierList ')'
+    Identifier = [a-zA-Z_$] [a-zA-Z_0-9]*
+    IdentifierList = Identifier ( ',' Identifier)*
+    Literal =
+        NumberLiteral | StringLiteral | HexLiteral
+    NumberLiteral = HexNumber | DecimalNumber
+    HexLiteral = 'hex' ('"' ([0-9a-fA-F]{2})* '"' | '\'' ([0-9a-fA-F]{2})* '\'')
+    StringLiteral = '"' ([^"\r\n\\] | '\\' .)* '"'
+    HexNumber = '0x' [0-9a-fA-F]+
+    DecimalNumber = [0-9]+
+
+Restrictions on the Grammar
+---------------------------
+
+Scopes in JULIA are tied to Blocks and all declarations
+(``FunctionDefinition``, ``VariableDeclaration`` and ``SubAssembly``)
+introduce new identifiers into these scopes. Shadowing is disallowed
+
+Talk about identifiers across functions etc
+
+
+Restriction for Expression: Statements have to return empty tuple
+Function arguments have to be single item
+
+Restriction for VariableDeclaration and Assignment: Number of elements left and right needs to be the same
+continue and break only in for loop
+
+Literals have to fit 32 bytes
+
+
+
+     | 'dataSize' '(' Identifier ')' |
+        LinkerSymbol |
+        'bytecodeSize' |
+
+
+Formal Specification
+--------------------
+
+We formally specify JULIA by providing an evaluation function E overloaded
+on the various nodes of the AST. Any functions can have side effects, so
+E takes a state objects and the actual argument and also returns new
+state objects and new arguments. There is a global state object
+(which in the context of the EVM is the memory, storage and state of the
+blockchain) and a local state object (the state of local variables, i.e. a
+segment of the stack in the EVM).
+
+The the evaluation function E takes a global state, a local state and
+a node of the AST and returns a new global state, a new local state
+and a value (if the AST node is an expression).
+
+We use sequence numbers as a shorthand for the order of evaluation
+and how state is forwarded. For example, ``E2(x), E1(y)`` is a shorthand
+for
+
+For ``(S1, z) = E(S, y)`` let ``(S2, w) = E(S1, x)``. TODO
+
+.. code::
+    E(G, L, <{St1, ..., Stn}>: Block) =
+        let L' be a copy of L that adds a new inner scope which contains
+        all functions and variables declared in the block (but not its sub-blocks)
+        variables are marked inactive for now
+        TODO: more formal
+        G1, L'1 = E(G, L', St1)
+        G2, L'2 = E(G1, L'1, St2)
+        ...
+        Gn, L'n = E(G(n-1), L'(n-1), Stn)
+        let L'' be a copy of L'n where the innermost scope is removed
+        Gn, L''
+    E(G, L, <function fname (param1, ..., paramn) -> (ret1, ..., retm) block>: FunctionDefinition) =
+        G, L
```

Comment:
Maybe not here, but somewhere, the function definition needs to be registered in `L`.  Not just the name of the function, but also the parameters, return values and the body.